### PR TITLE
Password check Homework

### PR DESCRIPTION
**Pull request by runer53248**

### CODE REVIEW

**Comment by ziobron** [comment]
✅ CI Passed - Task: password-check
🏅 Granted 8 XP for @runer53248

**Comment by ziobron** [comment]
✅ CI Passed - Task: password-check-bonus
🏅 Granted 4 XP for @runer53248

File: homework/password-check/validation.cpp [hunk 1]
```diff
@@ -1,2 +1,73 @@
...
+const std::map<ErrorCode, const std::string> errorsText = {
+    {ErrorCode::Ok, "Ok"},
+    {ErrorCode::PasswordNeedsAtLeastNineCharacters, "Password needs to have at least nine characters"},
+    {ErrorCode::PasswordNeedsAtLeastOneNumber, "Password needs to have at least one number"},
+    {ErrorCode::PasswordNeedsAtLeastOneSpecialCharacter, "Password needs to have at least one special character"},
+    {ErrorCode::PasswordNeedsAtLeastOneUppercaseLetter, "Password needs to have at least one uppercase letter"},
+    {ErrorCode::PasswordsDoNotMatch, "Passwords do not match"}};
+
+const std::string& getErrorMessage(const ErrorCode error) {
+    return errorsText.at(error);
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:17]: if you use .at you also need to handle exception :)

File: homework/password-check/validation.cpp [hunk 2]
```diff
@@ -1,2 +1,73 @@
...
+
+constexpr unsigned int letters_minimum_number = 9;
+
+int countLetters(int) {
+    static unsigned int lenght = 0;
+    if (++lenght == letters_minimum_number) {
+        lenght = 0;
+        return true;
+    }
+    return false;
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:32]: Don't use static here. Just use std::count, or count it inside for loop

File: homework/password-check/validation.cpp [hunk 3]
```diff
@@ -1,2 +1,73 @@
...
+    int (&test)(int);
+    const ErrorCode error;
+
+    constexpr Rule(int (&test)(int), ErrorCode error)
+        : test(test), error(error) {}
+
+    std::optional<ErrorCode> check(const std::string& password) const {
+        bool fail = std::none_of(begin(password), end(password), [=](char letter) {
+            return test(letter);
+        });
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:45]: you can check it inside if

File: homework/password-check/validation.cpp [hunk 4]
```diff
@@ -1,2 +1,73 @@
...
+        if (fail) {
+            return error;
+        }
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:48]: redundant

File: homework/password-check/validation.cpp [hunk 5]
```diff
@@ -1,2 +1,73 @@
...
+        return std::nullopt;
+    }
+};
+
+constexpr std::array rules{
+    Rule{countLetters, ErrorCode::PasswordNeedsAtLeastNineCharacters},
+    Rule{::isupper, ErrorCode::PasswordNeedsAtLeastOneUppercaseLetter},
+    Rule{::isdigit, ErrorCode::PasswordNeedsAtLeastOneNumber},
+    Rule{::ispunct, ErrorCode::PasswordNeedsAtLeastOneSpecialCharacter}};
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:57]: Nice idea

**Review by nauka-programowania-MA — COMMENTED** [review]
Really interesting solution

**Pull request closed without merging**